src/main/java/org/opentripplanner/transit/raptor/api/transit/RaptorTripSchedule.java

```diff
@@ -114,6 +116,40 @@ default int findDepartureStopPosition(int earliestDepartureTime, int stop) {
     return p.findStopPositionAfter(i, stop);
   }
 
+  /**
+   * Find all departure-stop-positions for a stop index after given earliest-departure-time. This is
+   * useful because trip can pass through the same stop more than once.
+   * This is basically on upgraded version of {@link #findDepartureStopPosition} that does not return
+   * only first stop position but all of them.
+   * @param earliestDepartureTime earliest departure time
+   * @param stop stop index
+   * @return list of all valid stop positions for a given stop index
+   */
+  default List<Integer> findStopPositionsForStopIndex(int earliestDepartureTime, int stop) {
```

Comment:
What are the performance implications of searching through the entire schedule? Is it worthwhile to generate a separate implementation for cases where we have detected a circular route?

I'm asking because the vast majority of routes are not circular and always doing some extra calculations for a rare case might be overkill.

But perhaps it's not really noticeable.

Reply:
Haven't performed any tests on it yet but my guess is that it should not really be that big performance hit. It's only done in OptimizeTransfers so in most cases there will only be handful of trip patterns it has to loop through. It's not doing any complex operations either, just comparing two integers and including them into a list. 

Theoretically we could calculate during graph-build stage whether trip pattern is circular or not but I'm not sure whether it would be worth it.

Reply:
That's fine. I think Thomas and Hannes will have a more detailed understanding of where the performance pitfalls of this code are.

Reply:
We could probably optimize this by including flag for circular-trips, that might be useful in other cases as well. But, I think this is ok for now, as @Bartosz-Kruba say, this is not used in Raptor - only in Optimized transfers.

But @leonardehrenfried you are right, changing anything in the `org.opentripplanner.transit.raptor` package is subject to extra caution.  This is not used inside Raptor; hence should not be added here. The method should be in another place, but that might be difficult. The optimized transfers operate on the Raptor interface - witch is not correct, but that is outside the scope of this PR, so for know this is ok. I might move this code to another place later.

Reply:
Maybe rename the method from `findStopPositionsForStopIndex` to `findDepartureStopPositions`. We search for both departures and arrivals, so that need to be part of the name, while I generally find it not so good to repeat the arguments names.

Reply:
I've changed the name now